test/plugins/test_substitute.py

```diff
@@ -0,0 +1,116 @@
+# This file is part of beets.
+# Copyright 2024, Nicholas Boyd Isacsson.
+#
+# Permission is hereby granted, free of charge, to any person obtaining
+# a copy of this software and associated documentation files (the
+# "Software"), to deal in the Software without restriction, including
+# without limitation the rights to use, copy, modify, merge, publish,
+# distribute, sublicense, and/or sell copies of the Software, and to
+# permit persons to whom the Software is furnished to do so, subject to
+# the following conditions:
+#
+# The above copyright notice and this permission notice shall be
+# included in all copies or substantial portions of the Software.
+
+"""Test the substitute plugin regex functionality."""
+
+from beets.test.helper import PluginTestCase, capture_log
+from beetsplug.substitute import Substitute
+
+PLUGIN_NAME = "substitute"
+
+
+class SubstitutePluginTest(PluginTestCase):
+    plugin = "substitute"
+    preload_plugin = False
+
+    def test_simple_substitute(self):
+        with self.configure_plugin(
+            [
+                {"a": "b"},
+                {"b": "c"},
+                {"c": "d"},
+            ]
+        ):
+            cases = [("a", "b"), ("b", "c"), ("c", "d")]
+            for input, expected in cases:
+                assert Substitute().tmpl_substitute(input) == expected
+
+    def test_case_insensitivity(self):
+        with self.configure_plugin([{"a": "b"}]):
+            assert Substitute().tmpl_substitute("A") == "b"
+
+    def test_unmatched_input_preserved(self):
+        with self.configure_plugin([{"a": "b"}]):
+            assert Substitute().tmpl_substitute("c") == "c"
```

Comment:
I think you can remove some duplication found in each of the test cases by defining a `run_substitute` method which accepts the configuration under test and the test cases. I 

```python
class SubstitutePluginTest(PluginTestCase):
    plugin = "substitute"
    preload_plugin = False

    def run_substitute(self, config, cases):
        with self.configure_plugin(config):
            for input, expected in cases:
                assert Substitute().tmpl_substitute(input) == expected

    def test_simple_substitute(self):
        self.run_substitute(
            [
                {"a": "b"},
                {"b": "c"},
                {"c": "d"},
            ],
            [("a", "b"), ("b", "c"), ("c", "d")],
        )

    def test_case_insensitivity(self):
        self.run_substitute([{"a": "b"}], [("A", "b")])

    def test_unmatched_input_preserved(self):
        self.run_substitute([{"a": "b"}], [("c", "c")])

    ...
```

Reply:
This simplified things a lot, many thanks!